src/zmalloc.c

```diff
@@ -361,22 +361,33 @@ size_t zmalloc_usable_size(void *ptr) {
 }
 #endif
 
-void zfree(void *ptr) {
-#ifndef HAVE_MALLOC_SIZE
-    void *realptr;
-    size_t oldsize;
+/* Frees the memory buffer pointed by ptr and updates statistics.
+ * ptr must already point to the start of the buffer. On systems where we store
+ * an additional header, the caller must do the necessary adjustments.
```

Comment:
what does this sentence means exactly? isn't it that the user must take full ownership on providing the REAL allocation size?

Reply:
If we don't have malloc_size on the system, we keep an extra header with the buffer size. In this case, the caller should adjust the pointer, to point to the start of the buffer. Not only to the the start of the logical object.
And, of course, the caller is responsible to provide the real allocation size.
If the caller can't satisfy any of the 2 conditions above, it should use `zfree` instead.

Reply:
I thin the documentation does not clearly capture all that. can you please consider refactor to provide a more detailed explanation?

Reply:
Updated.